We have use the `nltk.corpus` `movie_reviews` to get the data set



> From this we have seperate out the positive reviews data in `pos_reviews` and 
negative reviews data in `neg_reviews`



---







In [0]:
from nltk.corpus import movie_reviews 
import nltk

nltk.download('movie_reviews') # downloading the data set

pos_reviews = [] 
for ids in movie_reviews.fileids('pos'):
    words = movie_reviews.words(ids)
    pos_reviews.append(words)
 
neg_reviews = []
for ids in movie_reviews.fileids('neg'):
    words = movie_reviews.words(ids)
    neg_reviews.append(words)
 
# print first positive review item from the pos_reviews list
print ("pos_reviews[0] :", pos_reviews[0])
print()
# print first negative review item from the neg_reviews list
print ("neg_reviews[0] :", neg_reviews[0])
print()
# print first 20 items of the first item of positive review
print ("pos_reviews[0][:20] :", pos_reviews[0][:20])
print()
# print first 20 items of the first item of negative review
print ("neg_reviews[0][:20] :", neg_reviews[0][:20])


pos_reviews[0] : ['films', 'adapted', 'from', 'comic', 'books', 'have', ...]

neg_reviews[0] : ['plot', ':', 'two', 'teen', 'couples', 'go', 'to', ...]

pos_reviews[0][:20] : ['films', 'adapted', 'from', 'comic', 'books', 'have', 'had', 'plenty', 'of', 'success', ',', 'whether', 'they', "'", 're', 'about', 'superheroes', '(', 'batman', ',']

neg_reviews[0][:20] : ['plot', ':', 'two', 'teen', 'couples', 'go', 'to', 'a', 'church', 'party', ',', 'drink', 'and', 'then', 'drive', '.', 'they', 'get', 'into', 'an']


[nltk_data] Downloading package movie_reviews to
[nltk_data]     /Users/apple/nltk_data...
[nltk_data]   Package movie_reviews is already up-to-date!


Firstly, We will remove stop words and punctuations.

 # why?
> Because,
**Stop words** are those frequently words which do not carry any significant meaning in text analysis. For example, I, me, my, the, a, and, is, are, he, she, we, etc. **Punctuation marks** like comma, fullstop. inverted comma, etc. occur highly in any text data.

# Create Feature Set
<ul><li>Now, we write a function that will be used to create feature set. The feature set is used to train the classifier.</li>
<li>We use the bag-of-words feature as mentioned in the README.md.</li></ul>

In [0]:
from nltk.corpus import stopwords 
import string 
nltk.download('stopwords')
_stopwords = stopwords.words('english') # stopwords of English language because our model will work on English language only
 
# feature extractor function
def bag_of_words(words):
    words_clean = []
 
    for word in words:
        word = word.lower()
        if (word not in _stopwords) and (word not in string.punctuation): # string.punctuation will return 
            words_clean.append(word)                                    # string containing all punctuations
            
    words_dictionary = dict([word, True] for word in words_clean)
    
    return words_dictionary
 
# using dict will remove duplicate words from the words list
# note the output: stopword 'the' is also removed
print ("output:\nbag_of_words(['the', 'the', 'bad', 'bad', 'the', 'good']) =", bag_of_words(['the', 'the', 'bad', 'bad', 'the', 'good']))

output:
bag_of_words(['the', 'the', 'bad', 'bad', 'the', 'good']) = {'bad': True, 'good': True}


[nltk_data] Downloading package stopwords to /Users/apple/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Here, we made the pos_reviews_set and neg_reviews_set which is different from the above defined `pos_reviews` and `neg_reviews`.
The `pos_reviews_set` and `neg_reviews_set` content the words with its sentiment.

which is use to create **train** and **test set**.

> **Train Set** : is used to train the classifier.

> **Test Set** : is used to test the classifier to check how accurately it classifies the given text. 


---



In [0]:
# positive reviews feature set
pos_reviews_set = []
for words in pos_reviews:
    pos_reviews_set.append((bag_of_words(words), 'pos'))
 
# negative reviews feature set
neg_reviews_set = []
for words in neg_reviews:
    neg_reviews_set.append((bag_of_words(words), 'neg'))
 
# print first positive review item from the pos_reviews list
print ("pos_reviews_set[0] : ", pos_reviews_set[0])
print()
# print first negative review item from the neg_reviews list
print ("neg_reviews_set[0] : ", neg_reviews_set[0])


pos_reviews_set[0] :  ({'films': True, 'adapted': True, 'comic': True, 'books': True, 'plenty': True, 'success': True, 'whether': True, 'superheroes': True, 'batman': True, 'superman': True, 'spawn': True, 'geared': True, 'toward': True, 'kids': True, 'casper': True, 'arthouse': True, 'crowd': True, 'ghost': True, 'world': True, 'never': True, 'really': True, 'book': True, 'like': True, 'hell': True, 'starters': True, 'created': True, 'alan': True, 'moore': True, 'eddie': True, 'campbell': True, 'brought': True, 'medium': True, 'whole': True, 'new': True, 'level': True, 'mid': True, '80s': True, '12': True, 'part': True, 'series': True, 'called': True, 'watchmen': True, 'say': True, 'thoroughly': True, 'researched': True, 'subject': True, 'jack': True, 'ripper': True, 'would': True, 'saying': True, 'michael': True, 'jackson': True, 'starting': True, 'look': True, 'little': True, 'odd': True, 'graphic': True, 'novel': True, '500': True, 'pages': True, 'long': True, 'includes': True, 'ne

<h3>Here we have just created the train set and test set with train_set and test_set arrays respectively.</h3>

# Create Train and Test Set


> There are 1000 positive reviews set and 1000 negative reviews set. We take 20% (i.e. 200) of positive reviews and 20% (i.e. 200) of negative reviews as a test set. The remaining negative and positive reviews will be taken as a training set.


---


In [0]:
print ("len(pos_reviews_set), len(neg_reviews_set) :- " ,len(pos_reviews_set), len(neg_reviews_set)) # Output: (1000, 1000)
 
# radomize pos_reviews_set and neg_reviews_set
# doing so will output different accuracy result everytime we run the program
from random import shuffle 
shuffle(pos_reviews_set)
shuffle(neg_reviews_set)
 
test_set = pos_reviews_set[:200] + neg_reviews_set[:200]
train_set = pos_reviews_set[200:] + neg_reviews_set[200:]
X_count_vect =  pos_reviews_set + neg_reviews_set
print("len(test_set),  len(train_set) :- ", len(test_set),  len(train_set)) 

len(pos_reviews_set), len(neg_reviews_set) :-  1000 1000
len(test_set),  len(train_set) :-  400 1600


# Testing the trained Classifier

Let’s see the accuracy percentage of the trained classifier. The accuracy value changes each time you run the program because of the names array being shuffled above.

---

# Training Classifier and Calculating Accuracy

We train Naive Bayes Classifier using the training set and calculate the classification accuracy of the trained classifier using the test set.

 more on Naive Bayes Classifier:
> https://en.wikipedia.org/wiki/Naive_Bayes_classifier

---




In [0]:
from nltk import classify
from nltk import NaiveBayesClassifier
 
classifier = NaiveBayesClassifier.train(train_set)
 
accuracy = classify.accuracy(classifier, test_set)
print("accuracy :- ", accuracy) # Output: 0.7325
 
print (classifier.show_most_informative_features(10))

accuracy :-  0.68
Most Informative Features
                   anger = True              pos : neg    =     16.3 : 1.0
               insulting = True              neg : pos    =     13.7 : 1.0
               ludicrous = True              neg : pos    =     13.0 : 1.0
                    anna = True              pos : neg    =     11.0 : 1.0
                  avoids = True              pos : neg    =     11.0 : 1.0
              satisfying = True              pos : neg    =     11.0 : 1.0
             outstanding = True              pos : neg    =     10.6 : 1.0
                  regard = True              pos : neg    =     10.3 : 1.0
                seamless = True              pos : neg    =      9.7 : 1.0
                 rousing = True              pos : neg    =      9.7 : 1.0
None


# Observations:
Let’s see the most informative features among the entire features in the feature set.

The result shows that the word outstanding is used in positive reviews 10.6 times more often than it is used in negative reviews the word insulting is used in negative reviews 13.7 times more often than it is used in positive reviews. Similarly, for other letters. These ratios are also called likelihood ratios.

Therefore, a review has a high chance to be classified as positive if it contains words like outstanding and breathtaking. Similarly, a review has a high chance of being classified as negative if it contains words like insulting, awful, waste, etc.


---



# Testing Classifier with Custom Review

We provide custom review text and check the classification output of the trained classifier. The classifier correctly predicts both negative and positive reviews provided.


---



In [0]:
from nltk.tokenize import word_tokenize
nltk.download('punkt') 

custom_review = "Worst movie i have ever seen. It is worst then the corona virus, disasteres."
custom_review_tokens = word_tokenize(custom_review)
custom_review_set = bag_of_words(custom_review_tokens)
print ("Predicition on,\n'Worst movie i have ever seen. It is worst then the corona virus, disasteres.' custom reviews : ", classifier.classify(custom_review_set))
# Negative review correctly classified as negative
 
# probability result
prob_result = classifier.prob_classify(custom_review_set)
print (prob_result)
print (prob_result.max())
print (prob_result.prob("neg"))
print (prob_result.prob("pos"))
print()
custom_review = "Better then Corona, its good to watch movie like this, wonderful."
custom_review_tokens = word_tokenize(custom_review)
custom_review_set = bag_of_words(custom_review_tokens)
 
print ("Predicition on,\n'Better then Corona, its good to watch movie like this, wonderful.' custom reviews : ", classifier.classify(custom_review_set))
# Positive review correctly classified as positive
 
# probability result
prob_result = classifier.prob_classify(custom_review_set)
print (prob_result)
print (prob_result.max())
print (prob_result.prob("neg"))
print (prob_result.prob("pos"))

Predicition on,
'Worst movie i have ever seen. It is worst then the corona virus, disasteres.' custom reviews :  neg
<ProbDist with 2 samples>
neg
0.7122938651572529
0.2877061348427467

Predicition on,
'Better then Corona, its good to watch movie like this, wonderful.' custom reviews :  pos
<ProbDist with 2 samples>
pos
0.15550792749807166
0.844492072501929


[nltk_data] Downloading package punkt to /Users/apple/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
